src/rust/src/wallet_scanner.rs

```diff
@@ -0,0 +1,478 @@
+        self.acc.add_outputs(domain, outputs, tx);
+        self.pending_results.insert(txid, rx);
+
+        if self.acc.outputs.len() >= BATCH_SIZE_THRESHOLD {
+            self.flush();
+        }
+    }
+
+    /// Runs the currently accumulated batch on the global threadpool.
+    ///
+    /// Subsequent calls to `Self::add_outputs` will be accumulated into a new batch.
+    fn flush(&mut self) {
+        if !self.acc.is_empty() {
+            let mut batch = Batch::new(self.acc.ivks.clone());
+            mem::swap(&mut batch, &mut self.acc);
+            rayon::spawn_fifo(|| batch.run());
+        }
+    }
+
+    /// Collects the pending decryption results for the given transaction.
+    fn collect_results(&mut self, txid: TxId) -> HashMap<(TxId, usize), DecryptedNote<D>> {
+        self.pending_results
+            .remove(&txid)
+            // We won't have a pending result if the transaction didn't have outputs of
+            // this runner's kind.
+            .map(|rx| {
+                rx.into_iter()
+                    .filter_map(
+                        |OutputIndex {
+                             output_index,
+                             value,
+                         }| {
+                            value.map(|decrypted_note| ((txid, output_index), decrypted_note))
+                        },
+                    )
+                    .collect()
+            })
+            .unwrap_or_default()
+    }
+}
+
+/// A batch scanner for the `zcashd` wallet.
+struct BatchScanner {
+    params: Network,
+    sapling_runner: Option<BatchRunner<SaplingDomain<Network>, OutputDescription<GrothProofBytes>>>,
+}
+
+fn init_batch_scanner(
+    network: &Network,
+    sapling_ivks: &[[u8; 32]],
+) -> Result<Box<BatchScanner>, &'static str> {
+    let sapling_runner = if sapling_ivks.is_empty() {
+        None
+    } else {
+        let ivks = sapling_ivks
+            .iter()
+            .map(|ivk| {
+                let ivk: Option<sapling::SaplingIvk> =
+                    jubjub::Fr::from_bytes(ivk).map(sapling::SaplingIvk).into();
+                ivk.ok_or("Invalid Sapling ivk passed to wallet::init_batch_scanner()")
+            })
+            .collect::<Result<_, _>>()?;
+        Some(BatchRunner::new(ivks))
+    };
+
+    Ok(Box::new(BatchScanner {
+        params: *network,
+        sapling_runner,
+    }))
+}
+
+impl BatchScanner {
+    /// Adds the given transaction's shielded outputs to the various batch runners.
+    ///
+    /// After adding the outputs, any accumulated batch of sufficient size is run on the
+    /// global threadpool. Subsequent calls to `Self::add_transaction` will accumulate
+    /// those output kinds into new batches.
+    fn add_transaction(&mut self, tx_bytes: &[u8], height: u32) -> Result<(), io::Error> {
+        // The consensusBranchId parameter is ignored; it is not used in trial decryption.
+        let tx = Transaction::read(tx_bytes, consensus::BranchId::Sprout)?;
```

Comment:
I confirmed that the consensus branch ID is just copied into the `data: TransactionData` field of the `Transaction` and does not affect parsing (for v4; it isn't used at all for v5). https://github.com/zcash/librustzcash/blob/37fc28634e811fdf8db9274a0080cff17c6b6a09/zcash_primitives/src/transaction/mod.rs#L507-L618